scipy/optimize/meson.build

```diff
@@ -0,0 +1,191 @@
+  c_args : numpy_nodepr_api,
+  fortran_args : fortran_ignore_warnings,
+  include_directories : [inc_np, inc_f2py],
+  dependencies : [py3_dep],
+  install: true,
+  subdir: 'scipy/optimize')
+
+moduleTNC = py3.extension_module('moduleTNC',
+  ['tnc/tnc.h', 'tnc/moduleTNC.c', 'tnc/tnc.c'],
+  c_args: numpy_nodepr_api,
+  include_directories : [inc_np, inc_f2py],
+  dependencies : [py3_dep],
+  install: true,
+  subdir: 'scipy/optimize')
+
+cobyla_module = custom_target('cobyla_module',
+  output : ['_cobylamodule.c'],
+  input : 'cobyla/cobyla.pyf',
+  command : [py3, generate_f2pymod, '@INPUT@', '-o', '@OUTDIR@'],
+)
+
+_cobyla = py3.extension_module('_cobyla',
+  [cobyla_module, fortranobject_c, 'cobyla/cobyla2.f', 'cobyla/trstlp.f'],
+  c_args: numpy_nodepr_api,
+  fortran_args : fortran_ignore_warnings,
+  include_directories : [inc_np, inc_f2py],
+  dependencies : [py3_dep],
+  install: true,
+  subdir: 'scipy/optimize')
+
+minpack2_module = custom_target('minpack2_module',
+  output : ['minpack2module.c'],
+  input : 'minpack2/minpack2.pyf',
+  command : [py3, generate_f2pymod, '@INPUT@', '-o', '@OUTDIR@'],
+)
+
+minpack2 = py3.extension_module('minpack2',
+  [minpack2_module, fortranobject_c, 'minpack2/dcsrch.f', 'minpack2/dcstep.f'],
+  c_args: numpy_nodepr_api,
+  fortran_args : fortran_ignore_warnings,
+  include_directories : [inc_np, inc_f2py],
+  dependencies : [py3_dep],
+  install: true,
+  subdir: 'scipy/optimize')
+  
+slsqp_module = custom_target('slsqp_module',
+  output : ['_slsqpmodule.c'],
+  input : 'slsqp/slsqp.pyf',
+  command : [py3, generate_f2pymod, '@INPUT@', '-o', '@OUTDIR@'],
+)
+
+_slsqp = py3.extension_module('_slsqp',
+  [slsqp_module, fortranobject_c, 'slsqp/slsqp_optmz.f'],
+  c_args: numpy_nodepr_api,
+  fortran_args : fortran_ignore_warnings,
+  include_directories : [inc_np, inc_f2py],
+  dependencies : [py3_dep],
+  install: true,
+  subdir: 'scipy/optimize')
+
+__nnls_module = custom_target('__nnls_module',
+  output : ['__nnlsmodule.c'],
+  input : '__nnls/nnls.pyf',
+  command : [py3, generate_f2pymod, '@INPUT@', '-o', '@OUTDIR@'],
+)
+
+py3.install_sources(
+    ['__nnls.pyi'],
+    pure: false,
+    subdir: 'scipy/optimize'
+)
+
+__nnls = py3.extension_module('__nnls',
+  ['__nnls/nnls.f', __nnls_module, fortranobject_c],
+  c_args: numpy_nodepr_api,
+  fortran_args : fortran_ignore_warnings,
+  include_directories : [inc_np, inc_f2py],
+  dependencies : [py3_dep],
+  install: true,
+  subdir: 'scipy/optimize')
+
+res = run_command(py3, '-c', 'import os; print(os.environ.get("SCIPY_USE_PYTHRAN", 1))')
+if res.returncode() == 0
+  # DOES THIS WORK?
+  _group_columns = custom_target('_group_columns',
+    output: ['_group_columns.so'],
+    input : '_group_columns.py',
+    command : [pythran, '@INPUT@', '-o', '@OUTDIR@/_group_columns.so'],
+  )
+  _group_columns = py3.extension_module('_group_columns',
+  [_group_columns],
+  install: true,
+  subdir: 'scipy/optimize')
```

Comment:
I am not sure if this is a good idea. Actually, `pythran` generates a `cpp` file which wasn't getting compiled. So, I just generated the shared binary using `pythran` command and then added it as an extension module. It seems to work but not sure if it is correct or even something acceptable.

Reply:
That's indeed not a good idea. Pythran will use `setuptools` under the hood, so doing it like this would be mixing two build systems with different behavior. We want the whole build to use only Meson (good for consistency, cache use, and build time).

When this happens, I usually use `meson introspect build/ -i --targets > tmp.json` and then looks at that file to figure out what's wrong in the compile command.

Reply:
I think the command to use is `[pythran, '@INPUT@', '-E', '@OUTDIR@/_group_columns.cpp']`. With `output : '_group_columns.cpp'` and adding `py3_dep` to the `extension_module` call, I'd expect that to work.

Reply:
Ah that's below. Should be `.cpp` and `cpp_args` there.

Reply:
There is one file in `pythran` by the name, `pythonic/core.hpp` which I think is specific to `pythran`. If `py3_dep` works then that would be great but if it doesn't then we will be needing to add `pythonic/core.hpp` to dependencies or include directories. `ninja` doesn't recognize this file if we try to compile to `_group_columns.cpp`. So, I thought to by pass this `cpp` file altogether but yeah then it diminishes the purpose of using `meson` build system.

Reply:
Ah yes, this is the first Pythran extension. It generated standard C++ code, so all that should be needed is to add the `pythonic/` dir to the include dirs. 

You'll want to do this one level up in `scipy/meson.build`, similar to the f2py and numpy include dirs.

Reply:
Or to be more precise, since the generated includes look like `#include <pythonic/core.hpp>`, the directory one up from `pythonic/` should be used.

Reply:
It seems like, `pythran` doesn't has `get_include`,
```python
Python 3.9.4 | packaged by conda-forge | (default, May 10 2021, 22:13:33) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pythran
>>> pythran.get_include()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'pythran' has no attribute 'get_include'
>>> import numpy
>>> numpy.get_include()
'/home/czgdp1807ssd/anaconda3/envs/scipy-dev-main/lib/python3.9/site-packages/numpy/core/include'
>>> 
```

Reply:
Is there any other way to obtain the path to include directory of `pythran`.

Reply:
Looks like there's no built-in Python function, but looks like it's simply (not tested):
```python
incdir_numpy = run_command(py3,
          ['-c', 'import os; os.chdir(".."); import pythran; print(pythran.__file__)'],
          check : true
        ).stdout().strip()
```

Reply:
eh sorry, that's the file - so instead add `import os; os.dirname(pythran.__file__)`

Reply:
Suggested adding `get_include()` to Pythran in https://github.com/serge-sans-paille/pythran/issues/1807.

Reply:
Update, `get_include` is now merged into Pythran and we can use it once pythran `0.9.12` is released.